http-service-hyper/src/lib.rs

```diff
@@ -56,37 +54,40 @@ where
                 connection,
             })
         }
-            .boxed()
-            .compat()
+        .boxed()
+        .compat()
     }
 }
 
-impl<H> hyper::service::Service for WrapConnection<H>
+impl<H, > hyper::service::Service for WrapConnection<H, >
 where
     H: HttpService,
 {
     type ReqBody = hyper::Body;
     type ResBody = hyper::Body;
     type Error = std::io::Error;
-    type Future = Compat<BoxFuture<'static, Result<http::Response<hyper::Body>, Self::Error>>>;
+    type Future = Compat03As01<BoxFuture<'static, Result<http::Response<hyper::Body>, Self::Error>>>;
 
     fn call(&mut self, req: http::Request<hyper::Body>) -> Self::Future {
+        // Convert Request
         let error = std::io::Error::from(std::io::ErrorKind::Other);
-        let req = req.map(|hyper_body| {
-            let stream = Compat01As03::new(hyper_body).map(|c| match c {
-                Ok(chunk) => Ok(chunk.into_bytes()),
-                Err(e) => Err(std::io::Error::new(std::io::ErrorKind::Other, e)),
-            });
-            Body::from_stream(stream)
+        let req = req.map(|body| {
+            let body_stream = Compat01As03::new(body)
+                .map(|chunk| chunk.map(|chunk| chunk.to_vec()))
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e));
+            let body_reader = body_stream.into_async_read();
+            Body::from_reader(Box::new(body_reader))
```

Comment:
We don't need to box here if the `Unpin` bound of `from_reader` is removed.

Reply:
Great catch; thanks!